solr/core/src/test/org/apache/solr/schema/RankFieldTest.java

```diff
@@ -0,0 +1,261 @@
+        "sort", RANK_1 + " desc"), 400);
+  }
+  
+  @Ignore("We currently don't fail these kinds of requests with other field types")
+  public void testFacetFails() throws IOException {
+    assertU(adoc(
+        "id", "testFacetFails",
+        RANK_1, "1"
+        ));
+    assertU(commit());
+    assertQEx("Can't facet on rank field", req(
+        "q", "id:testFacetFails",
+        "facet", "true",
+        "facet.field", RANK_1), 400);
+  }
+  
+  public void testTermQuery() throws IOException {
+    assertU(adoc(
+        "id", "testTermQuery",
+        RANK_1, "1",
+        RANK_2, "1"
+        ));
+    assertU(adoc(
+        "id", "testTermQuery2",
+        RANK_1, "1"
+        ));
+    assertU(commit());
+    assertQ(req("q", RANK_1 + ":*"), "//*[@numFound='2']");
+    assertQ(req("q", RANK_1 + ":[* TO *]"), "//*[@numFound='2']");
+    assertQ(req("q", RANK_2 + ":*"), "//*[@numFound='1']");
+    assertQ(req("q", RANK_2 + ":[* TO *]"), "//*[@numFound='1']");
+    
+    assertQEx("Term queries not supported", req("q", RANK_1 + ":1"), 400);
+    assertQEx("Range queries not supported", req("q", RANK_1 + ":[1 TO 10]"), 400);
+  }
+  
+  
+  public void testResponseQuery() throws IOException {
+    assertU(adoc(
+        "id", "testResponseQuery",
+        RANK_1, "1"
+        ));
+    assertU(commit());
+    // Ignore requests to retrieve rank
+    assertQ(req("q", RANK_1 + ":*",
+        "fl", "id," + RANK_1),
+        "//*[@numFound='1']",
+        "count(//result/doc[1]/str)=1");
+  }
+  
+  public void testRankQParserQuery() throws IOException {
+    assertU(adoc(
+        "id", "1",
+        "str_field", "foo",
+        RANK_1, "1",
+        RANK_2, "2"
+        ));
+    assertU(adoc(
+        "id", "2",
+        "str_field", "foo",
+        RANK_1, "2",
+        RANK_2, "1"
+        ));
+    assertU(commit());
+    assertQ(req("q", "str_field:foo _query_:{!rank f='" + RANK_1 + "' function='log' scalingFactor='1'}"),
+        "//*[@numFound='2']",
+        "//result/doc[1]/str[@name='id'][.='2']",
+        "//result/doc[2]/str[@name='id'][.='1']");
+    
+    assertQ(req("q", "str_field:foo _query_:{!rank f='" + RANK_2 + "' function='log' scalingFactor='1'}"),
+        "//*[@numFound='2']",
+        "//result/doc[1]/str[@name='id'][.='1']",
+        "//result/doc[2]/str[@name='id'][.='2']");
+    
+    assertQ(req("q", "foo",
+        "defType", "dismax",
+        "qf", "str_field^10",
+        "bq", "{!rank f='" + RANK_1 + "' function='log' scalingFactor='1'}"
+        ),
+        "//*[@numFound='2']",
+        "//result/doc[1]/str[@name='id'][.='2']",
+        "//result/doc[2]/str[@name='id'][.='1']");
+    
+    assertQ(req("q", "foo",
+        "defType", "dismax",
+        "qf", "str_field^10",
+        "bq", "{!rank f='" + RANK_2 + "' function='log' scalingFactor='1'}"
+        ),
+        "//*[@numFound='2']",
+        "//result/doc[1]/str[@name='id'][.='1']",
+        "//result/doc[2]/str[@name='id'][.='2']");
+  }
```

Comment:
I manually checked the 'score' from these two runs to ensure that the score was changing! Would it make sense to add this as a test?

```
String rsp = h.query(req("q", "foo",
        "defType", "dismax",
        "qf", "str_field^10",
        "fl", "*,score"
    ));

    rsp = h.query(req("q", "foo",
        "defType", "dismax",
        "qf", "str_field^10",
        "fl", "*,score",
        "bq", "{!rank f='" + RANK_2 + "' function='log' scalingFactor='1'}"
    ));
```

Reply:
Also I was curious to see if `debugQuery=true` does the right thing ( it does! )

```


<?xml version="1.0" encoding="UTF-8"?>
<response>
   <lst name="responseHeader">
      <int name="status">0</int>
      <int name="QTime">161</int>
      <lst name="params">
         <str name="q">foo</str>
         <str name="defType">dismax</str>
         <str name="qf">str_field^10</str>
         <str name="fl">*,score,rank_1,rank_2</str>
         <str name="wt">xml</str>
         <str name="debugQuery">true</str>
         <str name="bq">{!rank f='rank_2' function='log' scalingFactor='1'}</str>
      </lst>
   </lst>
   <result name="response" numFound="2" start="0" maxScore="1.9273467" numFoundExact="true">
      <doc>
         <str name="id">1</str>
         <str name="str_field">foo</str>
         <float name="score">1.9273467</float>
      </doc>
      <doc>
         <str name="id">2</str>
         <str name="str_field">foo</str>
         <float name="score">1.5218816</float>
      </doc>
   </result>
   <lst name="debug">
      <str name="rawquerystring">foo</str>
      <str name="querystring">foo</str>
      <str name="parsedquery">+DisjunctionMaxQuery(((str_field:foo)^10.0)) () FeatureQuery(FeatureQuery(field=_rank_, feature=rank_2, function=LogFunction(scalingFactor=1.0)))</str>
      <str name="parsedquery_toString">+((str_field:foo)^10.0) () FeatureQuery(field=_rank_, feature=rank_2, function=LogFunction(scalingFactor=1.0))</str>
      <lst name="explain">
         <str>1.9273467 = sum of:
  0.82873434 = weight(str_field:foo in 0) [SchemaSimilarity], result of:
    0.82873434 = score(freq=1.0), computed as boost * idf * tf from:
      10.0 = boost
      0.18232156 = idf, computed as log(1 + (N - n + 0.5) / (n + 0.5)) from:
        2 = n, number of documents containing term
        2 = N, total number of documents with field
      0.45454544 = tf, computed as freq / (freq + k1 * (1 - b + b * dl / avgdl)) from:
        1.0 = freq, occurrences of term within document
        1.2 = k1, term saturation parameter
        0.75 = b, length normalization parameter
        1.0 = dl, length of field
        1.0 = avgdl, average length of field
  1.0986123 = Log function on the _rank_ field for the rank_2 feature, computed as w * log(a + S) from:
    1.0 = w, weight of this function
    1.0 = a, scaling factor
    2.0 = S, feature value</str>
         <str>1.5218816 = sum of:
  0.82873434 = weight(str_field:foo in 1) [SchemaSimilarity], result of:
    0.82873434 = score(freq=1.0), computed as boost * idf * tf from:
      10.0 = boost
      0.18232156 = idf, computed as log(1 + (N - n + 0.5) / (n + 0.5)) from:
        2 = n, number of documents containing term
        2 = N, total number of documents with field
      0.45454544 = tf, computed as freq / (freq + k1 * (1 - b + b * dl / avgdl)) from:
        1.0 = freq, occurrences of term within document
        1.2 = k1, term saturation parameter
        0.75 = b, length normalization parameter
        1.0 = dl, length of field
        1.0 = avgdl, average length of field
  0.6931472 = Log function on the _rank_ field for the rank_2 feature, computed as w * log(a + S) from:
    1.0 = w, weight of this function
    1.0 = a, scaling factor
    1.0 = S, feature value</str>
      </lst>
      <str name="QParser">DisMaxQParser</str>
      <null name="altquerystring" />
      <arr name="boost_queries">
         <str>{!rank f='rank_2' function='log' scalingFactor='1'}</str>
      </arr>
      <arr name="parsed_boost_queries">
         <str>FeatureQuery(FeatureQuery(field=_rank_, feature=rank_2, function=LogFunction(scalingFactor=1.0)))</str>
      </arr>
      <null name="boostfuncs" />
      <lst name="timing">
         <double name="time">160.0</double>
         <lst name="prepare">
            <double name="time">1.0</double>
            <lst name="query">
               <double name="time">1.0</double>
            </lst>
            <lst name="facet">
               <double name="time">0.0</double>
            </lst>
            <lst name="facet_module">
               <double name="time">0.0</double>
            </lst>
            <lst name="mlt">
               <double name="time">0.0</double>
            </lst>
            <lst name="highlight">
               <double name="time">0.0</double>
            </lst>
            <lst name="stats">
               <double name="time">0.0</double>
            </lst>
            <lst name="expand">
               <double name="time">0.0</double>
            </lst>
            <lst name="terms">
               <double name="time">0.0</double>
            </lst>
            <lst name="debug">
               <double name="time">0.0</double>
            </lst>
         </lst>
         <lst name="process">
            <double name="time">155.0</double>
            <lst name="query">
               <double name="time">3.0</double>
            </lst>
            <lst name="facet">
               <double name="time">0.0</double>
            </lst>
            <lst name="facet_module">
               <double name="time">0.0</double>
            </lst>
            <lst name="mlt">
               <double name="time">0.0</double>
            </lst>
            <lst name="highlight">
               <double name="time">0.0</double>
            </lst>
            <lst name="stats">
               <double name="time">0.0</double>
            </lst>
            <lst name="expand">
               <double name="time">0.0</double>
            </lst>
            <lst name="terms">
               <double name="time">0.0</double>
            </lst>
            <lst name="debug">
               <double name="time">152.0</double>
            </lst>
         </lst>
      </lst>
   </lst>
</response>
```

Reply:
Added the "score change" test